### PR TITLE
[GOBBLIN-793] Separate SpecSerDe from SpecCatalogs and add GsonSpecSerDe

### PR DESCRIPTION
**Pull request by jack-moseley**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-793

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

- Separate `SpecSerDe` from `SpecCatalog`s out to `JavaSpecSerDe`
- Add `GsonSpecSerDe` (for now only supports `FlowSpec` since a serializer/deserializer must be implemented for each `Spec`)
- Improve `SpecSerDeException` to instead be thrown inside serialize/deserialize methods

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Added `FlowSpecSerializationTest`

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by jack-moseley** [comment]
@autumnust can you take a look? thanks

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_serde/GsonSpecSerDe.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.gobblin.runtime.api.FlowSpec;
+import org.apache.gobblin.runtime.api.Spec;
+import org.apache.gobblin.runtime.api.SpecSerDe;
+import org.apache.gobblin.runtime.api.SpecSerDeException;
+
+
+/**
+ * {@link SpecSerDe} that serializes as Json using {@link Gson}. Note that currently only {@link FlowSpec}s are supported.
+ */
+public class GsonSpecSerDe implements SpecSerDe {
```
Comment (autumnust) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_serde/GsonSpecSerDe.java]: I don't think we could make it work for different implementations of `Spec` so shall we just name it as `GsonFlowSpecSerDe` ?

**Comment by autumnust** [comment]
+1 LGTM.

**Comment by jack-moseley** [comment]
@sv2000 please merge

**Pull request closed without merging**